src/funcs.c

```diff
@@ -496,17 +485,16 @@ static void LockFuncArgs(Obj func, Int narg, const Obj * args)
 
 #endif
 
-static ALWAYS_INLINE Obj DoExecFunc(Obj func, Int narg, const Obj *arg, Int isAtomic)
+static ALWAYS_INLINE Obj DoExecFunc(Obj func, Int narg, const Obj *arg)
 {
     Bag                 oldLvars;       /* old values bag                  */
-    REMEMBER_LOCKSTACK();
+    CHECK_RECURSION_BEFORE
+
 #ifdef HPCGAP
-    if (isAtomic)
+    REMEMBER_LOCKSTACK();
+    if (LCKS_FUNC(func))
         LockFuncArgs(func, narg, arg);
-#else
-    (void)isAtomic; // unused
 #endif
-    CHECK_RECURSION_BEFORE
```

Comment:
So to summarize the changes here: in the above, we moved `CHECK_RECURSION_BEFORE` to come first; then we call `REMEMBER_LOCKSTACK` (but now only for HPC-GAP; in non-HPC-GAP, it was a no-op anyway). And finally we check `LCKS_FUNC(func)` to see if `LockFuncArgs` should be called.